[Bug 20467] Fix multiple player crash

abrisan opened the pull request:
This is still in a very primitive stage

runrevmark committed:
[Bug 20467] Fixed bug causing crash when using more than one player on mac

runrevmark commented on engine/src/mac-av-player.mm:
```diff
@@ -36,6 +36,7 @@
 @interface com_runrev_livecode_MCAVFoundationPlayerObserver: NSObject
 {
     MCAVFoundationPlayer *m_av_player;
+    bool is_processing;
```
I'd suggest making this m_posted_currenttimechanged

runrevmark commented on engine/src/mac-av-player.mm:
```diff
@@ -176,7 +179,7 @@ - (id)initWithPlayer: (MCAVFoundationPlayer *)player
     
     m_av_player = player;
-    
+    is_processing = NO;
```
Its a 'bool' so use 'true' and 'false' (YES / NO is for the obj-c BOOL type)

runrevmark commented on engine/src/mac-av-player.mm:
```diff
@@ -216,6 +219,21 @@ - (void)updateCurrentFrame
+- (void) doHandleCurrentTimeChanged
+{
+    m_av_player->HandleCurrentTimeChanged();
+    self -> is_processing = NO;
```
Set the flag *before* HandleCurrentTimeChanged is called. HandleCurrentTimeChanged does engine stuff, which might call wait, which might mean there's another update which occurs whilst it is processing - *after* the engine code has got the current time.

runrevmark commented on engine/src/mac-av-player.mm:
```diff
@@ -437,8 +455,12 @@ - (void)setPlayer:(AVPlayer *)player
+
 void MCAVFoundationPlayer::HandleCurrentTimeChanged(void)
 {
+    MCAssert(pthread_main_np() != 0);
```
This assert should go in:
```

void MCPlatformCallbackSendPlayerFrameChanged(MCPlatformPlayerRef player);
void MCPlatformCallbackSendPlayerMarkerChanged(MCPlatformPlayerRef player, MCPlatformPlayerDuration time);
void MCPlatformCallbackSendPlayerCurrentTimeChanged(MCPlatformPlayerRef player);
void MCPlatformCallbackSendPlayerFinished(MCPlatformPlayerRef player);
void MCPlatformCallbackSendPlayerBufferUpdated(MCPlatformPlayerRef player);
```

abrisan replied:
should be done

runrevmark replied:
Since the assert is now in a more appropriate place, this assert can be removed.

runrevmark commented on engine/src/platform-internal.h:
```diff
@@ -369,6 +369,10 @@ void MCPlatformCallbackSendMenuSelect(MCPlatformMenuRef menu, uindex_t item);
 
 void MCPlatformCallbackSendViewFocusSwitched(MCPlatformWindowRef window, uint32_t view_id);
 
+#ifndef _PTHREAD_H
```
You only need to include this in platform.cpp

abrisan replied:
@runrevmark  done

runrevmark commented on engine/src/mac-av-player.mm:
```diff
@@ -216,6 +217,12 @@ - (void)updateCurrentFrame
     m_av_player -> DoUpdateCurrentFrame(m_av_player);
 }
 
+- (void)handleCurrentTimeChanged
```
I don't think we need these changes anymore.

runrevmark commented on engine/src/mac-av-player.mm:
```diff
@@ -90,6 +90,7 @@ - (void)setPlayer:(AVPlayer *)player;
     AVPlayer *getPlayer(void);
     
     static void DoUpdateCurrentFrame(void *ctxt);
+    void HandleCurrentTimeChanged(void);
```
I think this can go back to being protected

abrisan commented:
@runrevmark updated

runrevmark commented on engine/src/mac-av-player.mm:
```diff
@@ -94,7 +94,7 @@ - (void)setPlayer:(AVPlayer *)player;
 	virtual void Realize(void);
 	virtual void Unrealize(void);
-    
+    void HandleCurrentTimeChanged(void);
```
I think this can go back to being private

runrevmark commented on engine/src/mac-av-player.mm:
```diff
@@ -544,17 +546,16 @@ - (void)setPlayer:(AVPlayer *)player
-    }
-    
-    CMTime t_output_item_time = [t_self -> m_player_item_video_output itemTimeForCVTimeStamp:*inOutputTime];
+    CMTime t_output_item_time;
```
Init this with 0, as it will stop an potential error on some compilers as it is used later on

runrevmark commented:
This is almost there - I suggest merging the commits together with a more suitable commit message... And adding a release note of course :)

runrevmark commented on engine/src/mac-av-player.mm:
```diff
@@ -437,6 +435,8 @@ - (void)setPlayer:(AVPlayer *)player
     }
 }
 
+#include <pthread.h>
```
I don't think you need this include anymore

runrevmark commented on engine/src/mac-av-player.mm:
```diff
@@ -162,7 +160,8 @@ static CVReturn MyDisplayLinkCallback (CVDisplayLinkRef displayLink,
     bool m_mirrored : 1;
-
+    
+    void HandleCurrentTimeChanged(void);
```
Probably best to move this back to where it was to eliminate the unnecessary churn in the source.

runrevmark commented:
@livecode-vulcan review ok 3fa6e6a

livecode-vulcan commented:
💙 review by @runrevmark ok 3fa6e6a59f8443b84f19b358574154b4720546bb

livecode-vulcan commented:
😞 test failure 3fa6e6a59f8443b84f19b358574154b4720546bb

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/926)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/942)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/916)
* try-community-universal-ios-iphoneos11.0: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.0/builds/91)
* try-community-universal-ios-iphonesimulator11.0: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.0/builds/92)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/920)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/942)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/926)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/942)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/925)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/935)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/957)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/921)

livecodeali commented:
@livecode-vulcan retry 3fa6e6a

livecode-vulcan commented:
😞 test failure 3fa6e6a59f8443b84f19b358574154b4720546bb

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/931)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/947)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/921)
* try-community-universal-ios-iphoneos11.0: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.0/builds/96)
* try-community-universal-ios-iphonesimulator11.0: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.0/builds/97)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/925)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/947)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/931)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/947)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/930)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/940)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/962)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/926)

livecodepanos commented:
This fails on Windows:

`..\..\..\engine\src\platform.cpp(393): fatal error C1083: Cannot open include file: 'pthread.h': No such file or directory [C:\vulcan\Community\worker\try-community-x86-win32\worktree\build-win-x86\livecode\engine\kernel.vcxproj]`

runrevmark committed:
[[ Bug 20479 ]] Don't use pthread.h on Windows

This patch ensures that the 'main thread' assertions added to the
player callbacks in debug mode aren't used on Windows where
there are no pthreads, and it isn't a problem.

livecodeali commented:
@livecode-vulcan review ok f2ee0a9
